v7.0.0-rc.1 requires Firebird 4.0+ headers - cannot compile with Firebird 3.0 client

## Problem

When attempting to compile php-firebird v7.0.0-rc.1 with Firebird 3.0 client libraries (installed via IBSurgeon `fb_vanilla-30.sh`), the build fails with:

```
/tmp/php-firebird/src/cpp/fb_array.hpp:35:10: fatal error: firebird/impl/blr.h: No such file or directory
   35 | #include <firebird/impl/blr.h>
      |          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make: *** [Makefile:251: firebird_utils.lo] Error 1
```

## Environment

- PHP: 8.1-8.5 (tested on GitHub Actions with `php:X.Y-cli-bookworm`)
- Firebird Server: 3.0 (also tested with 2.5, 4.0, 5.0)
- Firebird Client: Installed via IBSurgeon scripts (`fb_vanilla-30.sh`)
- OS: Debian Bookworm (Ubuntu-latest runners)

## Analysis

The `firebird/impl/blr.h` header file is part of the **Firebird 4.0+** client development headers. The Firebird 3.0 client installation does not include this subdirectory structure.

Firebird 3.0 headers are typically at:
- `/opt/firebird/include/ibase.h`
- `/opt/firebird/include/iberror.h`

Firebird 4.0+ headers include additional paths:
- `/opt/firebird/include/firebird/impl/blr.h`
- `/opt/firebird/include/firebird/impl/sql.h`

## Expected Behavior

php-firebird v7.0.0 should compile successfully using **Firebird 3.0 client libraries**, since:
1. Many production systems still run Firebird 3.0
2. The PHP extension should support connecting to all supported Firebird versions (2.5, 3.0, 4.0, 5.0)
3. The previous v6.2.0 compiles successfully against Firebird 3.0 headers

## Proposed Solutions

1. **Add conditional compilation** to support both Firebird 3.x and 4.x+ headers
2. **Provide fallback definitions** for `blr.h` types when the header is missing
3. **Document minimum Firebird client version** if v7.0.0 intentionally drops Firebird 3.0 client support

## Workaround

Currently using v6.2.0 which compiles against all Firebird versions.

## References

- doctrine-firebird-driver CI failure: https://github.com/satwareAG/doctrine-firebird-driver/actions
- IBSurgeon install scripts: https://github.com/IBSurgeon/firebirdlinuxinstall

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -525,8 +525,10 @@ jobs:
           # tests/config.inc reads FIREBIRD_HOST and uses it as the host
           # portion of connection strings; the database path is appended in
           # tests/firebird.inc. Since the Firebird server runs in the same
-          # container, we connect to localhost on the default port.
-          FIREBIRD_HOST: "localhost"
+          # container, we connect via IPv4 loopback (127.0.0.1) to avoid
+          # IPv6 resolution issues where "localhost" might resolve to ::1
+          # but Firebird only binds to 127.0.0.1.
+          FIREBIRD_HOST: "127.0.0.1"
           # Firebird server restricts database creation paths via DatabaseAccess.
           # Use /opt/firebird/data which is typically writable by the server.
           FIREBIRD_DB_DIR: "/opt/firebird/data"
@@ -564,7 +566,7 @@ jobs:
           echo "DEBUG: Testing connection..."
           php -d "extension=$(pwd)/modules/firebird.so" -r '
             echo "Extension loaded, testing service attach...\n";
-            if ($se = @fbird_service_attach("localhost", "SYSDBA", "masterkey")) {
+            if ($se = @fbird_service_attach("127.0.0.1", "SYSDBA", "masterkey")) {
               echo "SUCCESS: Connected to Firebird service manager\n";
               $ver = fbird_server_info($se, FBIRD_SVC_SERVER_VERSION);
               echo "Server version: $ver\n";

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -590,8 +590,5 @@ jobs:
 
           # Use make test which properly handles the PHP test runner setup
           # The TEST_PHP_ARGS environment is used by run-tests.php
-          make test TESTS=tests/ || {
-            echo "Some tests failed. Checking for skips vs failures..."
-            # Don't fail if only skips, fail if actual test failures
-            true
-          }
+          # IMPORTANT: Tests MUST fail the job if they fail - no error masking!
+          make test TESTS=tests/

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -86,19 +86,44 @@ jobs:
       # Install Firebird 3.0+ client libraries for compilation
       # ========================================================================
       # The extension requires Firebird 3.0+ client libraries (OO API) for
-      # compilation. We install FB 3.0 client first, then the target server
-      # version for testing. The FB 3.0+ client can connect to all server
-      # versions (2.5-5.0) via backward-compatible wire protocol.
+      # compilation. As of v7.0.0+, we support compilation against both
+      # Firebird 3.0 and 4.0+ client headers via fb_blr_compat.h.
+      # See: https://github.com/satwareAG/php-firebird/issues/19
       - name: Install Firebird 3.0+ client for compilation
+        env:
+          MATRIX_FIREBIRD_VERSION: ${{ matrix.firebird-version }}
         run: |
           set -eu
 
-          # Use Firebird 4.0 as client for compilation - it has the complete
-          # firebird/impl/blr.h header structure required by the extension.
-          # FB 4.0 client can still connect to all server versions (2.5-5.0).
-          FB_CLIENT_VERSION="4.0.6"
-          FB_CLIENT_TARBALL="Firebird-${FB_CLIENT_VERSION}.3221-0.amd64.tar.gz"
-          FB_CLIENT_URL="https://github.com/FirebirdSQL/firebird/releases/download/v${FB_CLIENT_VERSION}/${FB_CLIENT_TARBALL}"
+          # Select client version based on matrix - use same version for
+          # compilation and runtime for consistency.
+          # For FB 2.5 server testing, use FB 3.0 client (minimum supported).
+          case "${MATRIX_FIREBIRD_VERSION}" in
+            '2.5')
+              FB_CLIENT_VERSION="3.0.12"
+              FB_CLIENT_TARBALL="Firebird-${FB_CLIENT_VERSION}.33787-0.amd64.tar.gz"
+              FB_CLIENT_URL="https://github.com/FirebirdSQL/firebird/releases/download/v${FB_CLIENT_VERSION}/${FB_CLIENT_TARBALL}"
+              ;;
+            '3.0')
+              FB_CLIENT_VERSION="3.0.12"
+              FB_CLIENT_TARBALL="Firebird-${FB_CLIENT_VERSION}.33787-0.amd64.tar.gz"
+              FB_CLIENT_URL="https://github.com/FirebirdSQL/firebird/releases/download/v${FB_CLIENT_VERSION}/${FB_CLIENT_TARBALL}"
+              ;;
+            '4.0')
+              FB_CLIENT_VERSION="4.0.6"
+              FB_CLIENT_TARBALL="Firebird-${FB_CLIENT_VERSION}.3221-0.amd64.tar.gz"
+              FB_CLIENT_URL="https://github.com/FirebirdSQL/firebird/releases/download/v${FB_CLIENT_VERSION}/${FB_CLIENT_TARBALL}"
+              ;;
+            '5.0')
+              FB_CLIENT_VERSION="5.0.2"
+              FB_CLIENT_TARBALL="Firebird-${FB_CLIENT_VERSION}.1613-0-linux-x64.tar.gz"
+              FB_CLIENT_URL="https://github.com/FirebirdSQL/firebird/releases/download/v${FB_CLIENT_VERSION}/${FB_CLIENT_TARBALL}"
+              ;;
+            *)
+              echo "Unsupported Firebird version in matrix: ${MATRIX_FIREBIRD_VERSION}" >&2
+              exit 1
+              ;;
+          esac
           FB_CLIENT_DIR="/opt/firebird-client"
 
           echo "Installing Firebird ${FB_CLIENT_VERSION} client for compilation..."
@@ -201,18 +226,17 @@ jobs:
               cp -a opt/firebird/lib/* "${FB_CLIENT_DIR}/lib/"
             fi
 
-            # Verify firebird/impl/blr.h exists (required for compilation)
-            # FB 4.0+ includes this header in the proper location
+            # Check for firebird/impl/blr.h (only present in FB 4.0+)
+            # As of v7.0.0+, this header is optional - fb_blr_compat.h provides fallbacks
             echo "Header structure verification:"
-            ls -la "${FB_CLIENT_DIR}/include/firebird/" 2>/dev/null || echo "No firebird/ subdir"
-            ls -la "${FB_CLIENT_DIR}/include/firebird/impl/" 2>/dev/null || echo "No firebird/impl/ subdir"
+            ls -la "${FB_CLIENT_DIR}/include/firebird/" 2>/dev/null || echo "No firebird/ subdir (FB 3.0 - using fallback BLR definitions)"
+            ls -la "${FB_CLIENT_DIR}/include/firebird/impl/" 2>/dev/null || echo "No firebird/impl/ subdir (FB 3.0 - using fallback BLR definitions)"
 
-            if [ ! -f "${FB_CLIENT_DIR}/include/firebird/impl/blr.h" ]; then
-              echo "ERROR: firebird/impl/blr.h not found - FB 4.0 client should have it"
-              find "${FB_CLIENT_DIR}/include" -name "*.h" 2>/dev/null | head -30
-              exit 1
+            if [ -f "${FB_CLIENT_DIR}/include/firebird/impl/blr.h" ]; then
+              echo "INFO: firebird/impl/blr.h found (FB 4.0+ native header)"
+            else
+              echo "INFO: firebird/impl/blr.h not found (FB 3.0 - using fb_blr_compat.h fallback definitions)"
             fi
-            echo "SUCCESS: firebird/impl/blr.h found"
           else
             # Fallback: try direct structure
             echo "No buildroot.tar.gz found, trying direct structure..."

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -9,7 +9,16 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Fixed
 
+- **Firebird 3.0 runtime connection failure**: Fixed `CheckStatusWrapper::isDirty()` behavior difference between FB3 and FB4+. In FB3, `isDirty()` returns `true` even on successful operations (it means "status was touched"), causing false error detection. Changed all error checks in `fb_connection.hpp` to use `hasData()` which correctly checks for actual errors (STATE_ERRORS flag). The extension now successfully connects to FB 2.5 and FB 3.0 servers using the FB 3.0 client library.
+
 - **Firebird 3.0 compilation compatibility** (Issue #19): Added `fb_blr_compat.h` header with fallback BLR constant definitions for systems without `firebird/impl/blr.h`. The extension can now compile against Firebird 3.0 client libraries without requiring FB 4.0+ headers.
+
+### Added
+
+- **Local Docker development matrix** for Firebird 3.0 client testing:
+  - New `php84-fb3-dev` container with Firebird 3.0.12 client library
+  - Code quality scripts (`cppcheck.sh`, `clang_tidy.sh`, `generate_compdb.sh`) auto-detect Firebird installation location
+  - Documentation updated in `docs/development/docker.md`
 
 ### Changed
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,14 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Fixed
+
+- **Firebird 3.0 compilation compatibility** (Issue #19): Added `fb_blr_compat.h` header with fallback BLR constant definitions for systems without `firebird/impl/blr.h`. The extension can now compile against Firebird 3.0 client libraries without requiring FB 4.0+ headers.
+
+### Changed
+
+- GitHub Actions CI now tests with matching Firebird client version for each server version (FB 3.0 client for FB 2.5/3.0 servers, FB 4.0+ client for FB 4.0/5.0 servers).
+
 ## [7.0.0-rc.1] - 2025-12-21
 
 ### Fixed

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,7 +5,7 @@ All notable changes to the PHP Firebird Extension will be documented in this fil
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
-## [Unreleased]
+## [7.0.0-rc.2] - 2025-12-22
 
 ### Fixed
 
@@ -213,7 +213,8 @@ grep -r "ibase\." config/
 - [Upstream Issues Analysis](docs/UPSTREAM_ISSUE_ANALYSIS.md)
 - [Development History](docs/DEVELOPMENT_HISTORY.md)
 
-[Unreleased]: https://github.com/satwareAG/php-firebird/compare/v7.0.0-rc.1...HEAD
+[Unreleased]: https://github.com/satwareAG/php-firebird/compare/v7.0.0-rc.2...HEAD
+[7.0.0-rc.2]: https://github.com/satwareAG/php-firebird/compare/v7.0.0-rc.1...v7.0.0-rc.2
 [7.0.0-rc.1]: https://github.com/satwareAG/php-firebird/compare/v6.2.0...v7.0.0-rc.1
 [6.2.0]: https://github.com/satwareAG/php-firebird/releases/tag/v6.2.0
 [1.0.0]: https://github.com/satwareAG/php-firebird/releases/tag/v1.0.0

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -9,6 +9,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Fixed
 
+- **TIME type encoding corruption** (Issue #21): Fixed `fbird_parse_time()` using uninitialized struct fields when parsing standalone TIME values. The function checked `out->has_date` to skip initialization, but on stack-allocated structs this contained garbage values. When garbage was non-zero, initialization was skipped, leaving `fractions` with random data that corrupted TIME encoding (e.g., "10:10:10" became "22:56:29"). Now always initializes time fields regardless of struct state.
+
 - **Firebird 3.0 runtime connection failure**: Fixed `CheckStatusWrapper::isDirty()` behavior difference between FB3 and FB4+. In FB3, `isDirty()` returns `true` even on successful operations (it means "status was touched"), causing false error detection. Changed all error checks in `fb_connection.hpp` to use `hasData()` which correctly checks for actual errors (STATE_ERRORS flag). The extension now successfully connects to FB 2.5 and FB 3.0 servers using the FB 3.0 client library.
 
 - **Firebird 3.0 compilation compatibility** (Issue #19): Added `fb_blr_compat.h` header with fallback BLR constant definitions for systems without `firebird/impl/blr.h`. The extension can now compile against Firebird 3.0 client libraries without requiring FB 4.0+ headers.

--- a/docker/docker-compose.yml
+++ b/docker/docker-compose.yml
@@ -132,6 +132,33 @@ services:
       - firebird40
       - firebird50
 
+  # PHP 8.4 with Firebird 3.0 Client Library
+  # For testing extension compilation against Firebird 3.0 client
+  # Compatible with Firebird 2.5 and 3.0 servers
+  php84-fb3-dev:
+    build:
+      context: php
+      dockerfile: Dockerfile-8.4-fb3
+    volumes:
+      - ..:/ext
+      - fb30-data:/firebird
+    user: "${CURRENT_UID:-1000}:${CURRENT_GID:-1000}"
+    working_dir: /ext
+    command: tail -f /dev/null
+    environment:
+      - PHP_IDE_CONFIG=serverName=php-firebird-dev
+      - XDEBUG_CONFIG=client_host=host.docker.internal
+      - FIREBIRD_HOST=firebird30
+      - FIREBIRD_DB_DIR=/firebird
+      - FIREBIRD_HOME=/opt/firebird
+      - LD_LIBRARY_PATH=/opt/firebird/lib
+      - TZ=Europe/Berlin
+    extra_hosts:
+      - "host.docker.internal:host-gateway"
+    depends_on:
+      - firebird25
+      - firebird30
+
   # PHP 8.5 with Firebird 5.x Client Library
   # For testing extension against the latest Firebird client
   php85-fb5-dev:

--- a/docker/php/Dockerfile-8.4-fb3
+++ b/docker/php/Dockerfile-8.4-fb3
@@ -1,0 +1,75 @@
+# PHP 8.4 with Firebird 3.0 Client Library
+# For testing extension against Firebird 3.0 client (compatible with FB 2.5/3.0 servers)
+FROM php:8.4-cli AS base
+
+# Install development dependencies (without firebird-dev from apt)
+# FB 3.0 client requires ncurses and tommath libraries (create compat symlinks below)
+RUN apt-get update && apt-get install -y \
+    autoconf \
+    build-essential \
+    gcc \
+    g++ \
+    make \
+    libtool \
+    bison \
+    re2c \
+    libicu-dev \
+    libxml2-dev \
+    git \
+    unzip \
+    wget \
+    tzdata \
+    vim \
+    gdb \
+    valgrind \
+    lcov \
+    libncurses6 \
+    libtommath1 \
+    cppcheck \
+    clang-tidy \
+    bear \
+    python3-pip \
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/*
+
+# Create compatibility symlinks for FB 3.0 client dependencies
+RUN ln -sf /lib/x86_64-linux-gnu/libncurses.so.6 /lib/x86_64-linux-gnu/libncurses.so.5 2>/dev/null || true \
+    && ln -sf /usr/lib/x86_64-linux-gnu/libtommath.so.1 /usr/lib/x86_64-linux-gnu/libtommath.so.0 2>/dev/null || true \
+    && ldconfig
+
+# Download and install Firebird 3.0.12 client library
+# Note: FB 3.0 client is compatible with FB 2.5 and 3.0 servers
+# FB 3.0 uses .amd64.tar.gz naming (different from FB 5.0's -linux-x64.tar.gz)
+# Extracted directory is Firebird-3.0.12.33787-0.amd64/
+ARG FIREBIRD_VERSION=3.0.12.33787-0.amd64
+RUN cd /tmp \
+    && wget -q https://github.com/FirebirdSQL/firebird/releases/download/v3.0.12/Firebird-${FIREBIRD_VERSION}.tar.gz \
+    && tar xf Firebird-${FIREBIRD_VERSION}.tar.gz \
+    && mkdir -p /opt/firebird \
+    && tar xf Firebird-${FIREBIRD_VERSION}/buildroot.tar.gz -C /opt/firebird --strip-components=3 \
+    && rm -rf /tmp/Firebird-* \
+    && ldconfig
+
+# Create symlinks for the extension build system
+RUN ln -sf /opt/firebird/lib/libfbclient.so /usr/lib/libfbclient.so \
+    && ln -sf /opt/firebird/lib/libfbclient.so.2 /usr/lib/libfbclient.so.2 \
+    && ln -sf /opt/firebird/include /usr/include/firebird \
+    && echo "/opt/firebird/lib" > /etc/ld.so.conf.d/firebird.conf \
+    && ldconfig
+
+# Install pcntl extension for event handling tests
+RUN docker-php-ext-install pcntl
+
+# Set up entrypoint
+COPY entrypoint.sh /entrypoint.sh
+RUN chmod +x /entrypoint.sh
+
+# Environment variables for build and runtime
+# FIREBIRD: Used by client library to find message files (firebird.msg)
+# FIREBIRD_HOME: Custom variable for our build scripts
+ENV FIREBIRD=/opt/firebird
+ENV FIREBIRD_HOME=/opt/firebird
+ENV LD_LIBRARY_PATH=/opt/firebird/lib:$LD_LIBRARY_PATH
+
+ENTRYPOINT ["/entrypoint.sh"]
+CMD ["php", "-a"]

--- a/docs/development/docker.md
+++ b/docs/development/docker.md
@@ -44,11 +44,63 @@ The development environment includes Firebird database servers on these ports:
 - Firebird 2.5: `localhost:3050`
 - Firebird 3.0: `localhost:3051`
 - Firebird 4.0: `localhost:3052`
+- Firebird 5.0: `localhost:3053`
 
 Default credentials:
 - Username: `SYSDBA`
 - Password: `masterkey`
 - Database: `test.fdb`
+
+## Firebird Client Version Testing
+
+The development environment supports testing with different Firebird client library versions.
+
+### Standard Containers (Firebird 4.x Client)
+
+By default, PHP containers use the apt-installed Firebird 4.x client library:
+- `php81-dev`, `php82-dev`, `php83-dev`, `php84-dev`, `php85-dev`
+
+These containers can connect to all Firebird server versions (2.5, 3.0, 4.0, 5.0).
+
+### Firebird 3.0 Client Container
+
+For testing compilation against Firebird 3.0 client headers (Issue #19 compatibility):
+```bash
+docker compose exec php84-fb3-dev sh -c "cd /ext && phpize --clean && phpize && ./configure --with-firebird=/opt/firebird && make -j\$(nproc)"
+```
+
+The `php84-fb3-dev` container:
+- Uses FB 3.0.12 client library
+- Compatible with FB 2.5 and 3.0 servers
+- Firebird installed in `/opt/firebird`
+- Environment: `FIREBIRD_HOME=/opt/firebird`
+
+### Firebird 5.0 Client Container
+
+For testing with the latest Firebird 5.x client:
+```bash
+docker compose exec php85-fb5-dev sh -c "cd /ext && phpize --clean && phpize && ./configure --with-firebird=/opt/firebird && make -j\$(nproc)"
+```
+
+The `php85-fb5-dev` container:
+- Uses FB 5.0.3 client library
+- Best compatibility with FB 5.0 server features
+- Firebird installed in `/opt/firebird`
+- Environment: `FIREBIRD_HOME=/opt/firebird`
+
+### Code Quality Checks with Different Clients
+
+The static analysis scripts auto-detect the Firebird installation:
+```bash
+# Run cppcheck in FB 3.0 client container
+docker compose exec php84-fb3-dev sh -c "/ext/scripts/container/analysis/cppcheck.sh"
+
+# Run clang-tidy in FB 5.0 client container
+docker compose exec php85-fb5-dev sh -c "/ext/scripts/container/analysis/clang_tidy.sh"
+
+# Generate compile_commands.json (auto-detects Firebird path)
+docker compose exec php84-fb3-dev sh -c "/ext/scripts/container/analysis/generate_compdb.sh"
+```
 
 ## Customization
 

--- a/fbird_datetime.c
+++ b/fbird_datetime.c
@@ -336,11 +336,19 @@ int fbird_parse_date(const char* str, fbird_datetime_components* out) {
 int fbird_parse_time(const char* str, fbird_datetime_components* out) {
     if (!str || !out) return 0;
 
-    /* Don't reinitialize if called from fbird_parse_timestamp */
-    int init_needed = !out->has_date;
-    if (init_needed) {
-        fbird_datetime_init(out);
-    }
+    /* BUGFIX: Always initialize time-related fields to avoid garbage values.
+     * The previous logic checked out->has_date to skip init, but on uninitialized
+     * structs, has_date contains garbage which could skip initialization and leave
+     * fractions with random values, causing TIME encoding corruption (Issue #21).
+     *
+     * When called from fbird_parse_timestamp(), the date fields are already set,
+     * so we preserve them. We only zero-init the time fields. */
+    out->hours = 0;
+    out->minutes = 0;
+    out->seconds = 0;
+    out->fractions = 0;
+    out->has_time = 0;
+    /* Don't touch date fields or timezone - preserve if already set */
 
     /* Skip leading whitespace */
     while (*str && isspace((unsigned char)*str)) str++;

--- a/fbird_query_bind.c
+++ b/fbird_query_bind.c
@@ -697,29 +697,29 @@ int _php_fbird_bind(fbird_query *ib_query, zval *b_vars) /* {{{ */
 							break;
 					}
 
-					if (!parsed) {
-						/* Cross-platform parsing failed, let Firebird try as string */
-						break;
-					}
+				if (!parsed) {
+					/* Cross-platform parsing failed, let Firebird try as string */
+					break;
+				}
 
-					/* Encode using OO API with parsed components */
-					switch (var->sqltype & ~1) {
-						default: /* == case SQL_TIMESTAMP */
-							buf[i].val.tsval = fbu_encode_timestamp(IBG(master_instance),
-								dt.year, dt.month, dt.day,
-								dt.hours, dt.minutes, dt.seconds,
-								dt.fractions);
-							break;
-						case SQL_TYPE_DATE:
-							buf[i].val.dtval = fbu_encode_date(IBG(master_instance),
-								dt.year, dt.month, dt.day);
-							break;
-						case SQL_TYPE_TIME:
-							buf[i].val.tmval = fbu_encode_time(IBG(master_instance),
-								dt.hours, dt.minutes, dt.seconds,
-								dt.fractions);
-							break;
-					}
+				/* Encode using OO API with parsed components */
+				switch (var->sqltype & ~1) {
+					default: /* == case SQL_TIMESTAMP */
+						buf[i].val.tsval = fbu_encode_timestamp(IBG(master_instance),
+							dt.year, dt.month, dt.day,
+							dt.hours, dt.minutes, dt.seconds,
+							dt.fractions);
+						break;
+					case SQL_TYPE_DATE:
+						buf[i].val.dtval = fbu_encode_date(IBG(master_instance),
+							dt.year, dt.month, dt.day);
+						break;
+					case SQL_TYPE_TIME:
+						buf[i].val.tmval = fbu_encode_time(IBG(master_instance),
+							dt.hours, dt.minutes, dt.seconds,
+							dt.fractions);
+						break;
+				}
 				}
 				continue;
 

--- a/scripts/container/analysis/clang_tidy.sh
+++ b/scripts/container/analysis/clang_tidy.sh
@@ -70,6 +70,16 @@ if [ -f compile_commands.json ]; then
 else
     # Provide manual include paths if no compile_commands.json
     PHP_INCLUDE_DIR=$(php-config --include-dir 2>/dev/null || echo "/usr/include/php")
+
+    # Detect Firebird include path (supports both apt-installed and manual installations)
+    if [ -d "/opt/firebird/include" ]; then
+        FB_INCLUDE_DIR="/opt/firebird/include"
+    elif [ -d "/usr/include/firebird" ]; then
+        FB_INCLUDE_DIR="/usr/include/firebird"
+    else
+        FB_INCLUDE_DIR="/usr/include/firebird"
+    fi
+
     CLANG_TIDY_ARGS+=(
         "--"
         "-I."
@@ -77,11 +87,12 @@ else
         "-I${PHP_INCLUDE_DIR}/Zend"
         "-I${PHP_INCLUDE_DIR}/main"
         "-I${PHP_INCLUDE_DIR}/TSRM"
-        "-I/usr/include/firebird"
+        "-I${FB_INCLUDE_DIR}"
         "-std=c++17"
         "-DHAVE_CONFIG_H"
     )
     echo "Using manual include paths (compile_commands.json recommended)"
+    echo "Firebird include: ${FB_INCLUDE_DIR}"
 fi
 
 echo ""

--- a/scripts/container/analysis/generate_compdb.sh
+++ b/scripts/container/analysis/generate_compdb.sh
@@ -16,11 +16,27 @@ if [ -f Makefile ]; then
     phpize --clean 2>/dev/null || true
 fi
 
+# Detect Firebird installation location
+# Supports both apt-installed (/usr) and manual installations (/opt/firebird)
+if [ -d "/opt/firebird/include" ]; then
+    FB_HOME="/opt/firebird"
+    FB_INCLUDE="/opt/firebird/include"
+    echo "Detected manual Firebird installation at /opt/firebird"
+elif [ -d "/usr/include/firebird" ]; then
+    FB_HOME="/usr"
+    FB_INCLUDE="/usr/include/firebird"
+    echo "Detected apt-installed Firebird at /usr"
+else
+    FB_HOME="/usr"
+    FB_INCLUDE="/usr/include/firebird"
+    echo "Warning: Firebird headers not found, using default /usr"
+fi
+
 # Prepare build environment
 phpize
 
-# Configure with Firebird paths
-CPPFLAGS="-I/usr/include/firebird" ./configure --with-firebird=/usr
+# Configure with detected Firebird paths
+CPPFLAGS="-I${FB_INCLUDE}" ./configure --with-firebird="${FB_HOME}"
 
 # Generate compile_commands.json using bear (intercepts compilation commands)
 if command -v bear &> /dev/null; then

--- a/scripts/container/build.sh
+++ b/scripts/container/build.sh
@@ -17,11 +17,29 @@ fi
 # Prepare build environment
 phpize
 
+# Auto-detect Firebird paths
+# FIREBIRD_HOME environment variable takes precedence (set in FB3/FB5 containers)
+# Otherwise defaults to /usr (apt-installed firebird-dev package)
+if [ -n "$FIREBIRD_HOME" ] && [ -d "$FIREBIRD_HOME" ]; then
+    FIREBIRD_PATH="$FIREBIRD_HOME"
+    FIREBIRD_INCLUDE="$FIREBIRD_HOME/include"
+    echo "Using FIREBIRD_HOME: $FIREBIRD_PATH"
+elif [ -d "/opt/firebird" ]; then
+    FIREBIRD_PATH="/opt/firebird"
+    FIREBIRD_INCLUDE="/opt/firebird/include"
+    echo "Auto-detected Firebird at /opt/firebird"
+else
+    FIREBIRD_PATH="/usr"
+    FIREBIRD_INCLUDE="/usr/include/firebird"
+    echo "Using system Firebird at /usr"
+fi
+
 # Configure with Firebird paths
-CPPFLAGS="-I/usr/include/firebird" ./configure \
-    --with-firebird=/usr
+CPPFLAGS="-I$FIREBIRD_INCLUDE" ./configure \
+    --with-firebird="$FIREBIRD_PATH"
 
 # Build
 make -j$(nproc)
 
 echo "Build completed. Extension is available at: $(pwd)/modules/firebird.so"
+echo "Firebird client library: $FIREBIRD_PATH"

--- a/scripts/host/test_matrix.sh
+++ b/scripts/host/test_matrix.sh
@@ -51,7 +51,11 @@ fi
 if [ -n "$1" ]; then
     TARGETS=("$1")
 else
-    TARGETS=("php81-dev" "php82-dev" "php83-dev" "php84-dev" "php85-dev" "php85-fb5-dev")
+    # Test all PHP containers including special client library variants:
+    # - Standard containers (php8x-dev): Use apt firebird-dev (Firebird 4.x client)
+    # - php84-fb3-dev: Firebird 3.0.12 client for FB 2.5/3.0 server compatibility
+    # - php85-fb5-dev: Firebird 5.x client for latest features
+    TARGETS=("php81-dev" "php82-dev" "php83-dev" "php84-dev" "php84-fb3-dev" "php85-dev" "php85-fb5-dev")
 fi
 
 # 4. Parse Firebird Server Target (optional second argument)

--- a/src/cpp/fb_array.hpp
+++ b/src/cpp/fb_array.hpp
@@ -32,8 +32,7 @@
 #include <cstring>
 #include <cstdint>
 
-#include <firebird/impl/blr.h>
-
+#include "fb_blr_compat.h"
 #include "fb_status.hpp"
 
 // Charset ids for BLR *2 types (blr_text2/blr_varying2).

--- a/src/cpp/fb_blr_compat.h
+++ b/src/cpp/fb_blr_compat.h
@@ -1,0 +1,252 @@
+/* SPDX-License-Identifier: PHP-3.01
+ * SPDX-FileCopyrightText: The PHP Group and contributors (see CREDITS) */
+
+#ifndef FB_BLR_COMPAT_H
+#define FB_BLR_COMPAT_H
+
+/**
+ * @file fb_blr_compat.h
+ * @brief BLR (Binary Language Representation) constants compatibility header
+ *
+ * This header provides BLR constants needed for array operations.
+ *
+ * Firebird 4.0+ includes these in <firebird/impl/blr.h>
+ * Firebird 3.0 does NOT have this header - constants are internal.
+ *
+ * This compatibility header allows compilation with both Firebird 3.0 and 4.0+ client libraries.
+ *
+ * @see https://github.com/satwareAG/php-firebird/issues/19
+ */
+
+/*
+ * Try to include Firebird 4.0+ header if available.
+ * The __has_include() preprocessor feature is available in:
+ * - GCC 5+
+ * - Clang 3.0+
+ * - MSVC 2017+
+ *
+ * For compilers without __has_include, we fall through to fallback definitions.
+ */
+#ifdef __has_include
+#  if __has_include(<firebird/impl/blr.h>)
+#    include <firebird/impl/blr.h>
+#    define FB_BLR_COMPAT_HAVE_FB4_HEADER 1
+#  endif
+#endif
+
+/*
+ * Fallback BLR constant definitions for Firebird 3.0 compatibility.
+ *
+ * These values are stable across Firebird versions (2.5, 3.0, 4.0, 5.0).
+ * Source: Firebird source code src/include/firebird/impl/blr.h
+ *
+ * Only define if not already defined (either by Firebird 4.0+ header or ibase.h).
+ */
+
+/* Basic data types */
+#ifndef blr_text
+#define blr_text                14
+#endif
+
+#ifndef blr_text2
+#define blr_text2               15
+#endif
+
+#ifndef blr_short
+#define blr_short               7
+#endif
+
+#ifndef blr_long
+#define blr_long                8
+#endif
+
+#ifndef blr_quad
+#define blr_quad                9
+#endif
+
+#ifndef blr_float
+#define blr_float               10
+#endif
+
+#ifndef blr_double
+#define blr_double              27
+#endif
+
+#ifndef blr_d_float
+#define blr_d_float             11
+#endif
+
+#ifndef blr_timestamp
+#define blr_timestamp           35
+#endif
+
+#ifndef blr_varying
+#define blr_varying             37
+#endif
+
+#ifndef blr_varying2
+#define blr_varying2            38
+#endif
+
+#ifndef blr_blob
+#define blr_blob                261
+#endif
+
+#ifndef blr_cstring
+#define blr_cstring             40
+#endif
+
+#ifndef blr_cstring2
+#define blr_cstring2            41
+#endif
+
+#ifndef blr_blob_id
+#define blr_blob_id             45
+#endif
+
+#ifndef blr_sql_date
+#define blr_sql_date            12
+#endif
+
+#ifndef blr_sql_time
+#define blr_sql_time            13
+#endif
+
+#ifndef blr_int64
+#define blr_int64               16
+#endif
+
+/* Firebird 4.0+ types - needed for forward compatibility */
+#ifndef blr_int128
+#define blr_int128              26
+#endif
+
+#ifndef blr_dec64
+#define blr_dec64               24
+#endif
+
+#ifndef blr_dec128
+#define blr_dec128              25
+#endif
+
+#ifndef blr_sql_time_tz
+#define blr_sql_time_tz         28
+#endif
+
+#ifndef blr_timestamp_tz
+#define blr_timestamp_tz        29
+#endif
+
+#ifndef blr_ex_time_tz
+#define blr_ex_time_tz          30
+#endif
+
+#ifndef blr_ex_timestamp_tz
+#define blr_ex_timestamp_tz     31
+#endif
+
+#ifndef blr_bool
+#define blr_bool                23
+#endif
+
+/* BLR version and structural elements */
+#ifndef blr_version4
+#define blr_version4            4
+#endif
+
+#ifndef blr_version5
+#define blr_version5            5
+#endif
+
+#ifndef blr_eoc
+#define blr_eoc                 76
+#endif
+
+#ifndef blr_end
+#define blr_end                 255
+#endif
+
+/* Message and parameter definitions */
+#ifndef blr_message
+#define blr_message             4
+#endif
+
+#ifndef blr_begin
+#define blr_begin               2
+#endif
+
+#ifndef blr_send
+#define blr_send                3
+#endif
+
+#ifndef blr_receive
+#define blr_receive             6
+#endif
+
+#ifndef blr_loop
+#define blr_loop                17
+#endif
+
+#ifndef blr_for
+#define blr_for                 18
+#endif
+
+#ifndef blr_if
+#define blr_if                  19
+#endif
+
+#ifndef blr_stall
+#define blr_stall               20
+#endif
+
+#ifndef blr_select
+#define blr_select              21
+#endif
+
+/* Value fetch operations */
+#ifndef blr_field
+#define blr_field               50
+#endif
+
+#ifndef blr_literal
+#define blr_literal             51
+#endif
+
+#ifndef blr_dbkey
+#define blr_dbkey               52
+#endif
+
+#ifndef blr_parameter
+#define blr_parameter           53
+#endif
+
+#ifndef blr_parameter2
+#define blr_parameter2          54
+#endif
+
+#ifndef blr_null
+#define blr_null                55
+#endif
+
+/* Aggregate functions */
+#ifndef blr_agg_count
+#define blr_agg_count           60
+#endif
+
+#ifndef blr_agg_max
+#define blr_agg_max             61
+#endif
+
+#ifndef blr_agg_min
+#define blr_agg_min             62
+#endif
+
+#ifndef blr_agg_total
+#define blr_agg_total           63
+#endif
+
+#ifndef blr_agg_average
+#define blr_agg_average         64
+#endif
+
+#endif /* FB_BLR_COMPAT_H */

--- a/src/cpp/fb_connection.hpp
+++ b/src/cpp/fb_connection.hpp
@@ -382,7 +382,10 @@ inline Connection Connection::create(Firebird::IMaster* master,
         dpb.getBuffer()
     );
 
-    if (check_status.isDirty() || !raw_attachment) {
+    // Note: Use hasData() instead of isDirty() for FB3 compatibility.
+    // In FB3, isDirty() returns true even on success (it means "status was touched").
+    // hasData() correctly checks for actual errors (STATE_ERRORS flag).
+    if (check_status.hasData() || !raw_attachment) {
         throw Exception(raw_status);
     }
 
@@ -515,7 +518,8 @@ inline void Connection::detach() {
     Firebird::CheckStatusWrapper check_status(raw_status);
     attachment_->detach(&check_status);
 
-    if (check_status.isDirty()) {
+    // Use hasData() for FB3 compatibility (see Connection::create comment)
+    if (check_status.hasData()) {
         last_status_ = StatusWrapper(master_);
         throw Exception(raw_status);
     }
@@ -534,7 +538,8 @@ inline bool Connection::detachNoThrow() noexcept {
             Firebird::IStatus* raw_status = master_->getStatus();
             Firebird::CheckStatusWrapper check_status(raw_status);
             attachment_->detach(&check_status);
-            if (check_status.isDirty()) {
+            // Use hasData() for FB3 compatibility (see Connection::create comment)
+            if (check_status.hasData()) {
                 attachment_.reset();
                 return false;
             }
@@ -561,7 +566,8 @@ inline void Connection::dropDatabase() {
     Firebird::CheckStatusWrapper check_status(raw_status);
     attachment_->dropDatabase(&check_status);
 
-    if (check_status.isDirty()) {
+    // Use hasData() for FB3 compatibility (see Connection::create comment)
+    if (check_status.hasData()) {
         throw Exception(raw_status);
     }
 

